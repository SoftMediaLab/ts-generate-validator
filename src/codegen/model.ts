import { Message } from '../localization/model';
import { UserContext, ValidationConfig } from '../config/model';

export type Data = Record<string, any>;

export type GeneratedValidation = <D extends Data, C extends UserContext>(
  data: D,
  config?: ValidationConfig,
  context?: C
) => void | Promise<void>;

export interface ValidationError<D extends Data> {
  field: keyof D | (keyof D)[];
  message: Message;
}

export interface ValidationException<D extends Data> {
  errors: ValidationError<D>[];
}
