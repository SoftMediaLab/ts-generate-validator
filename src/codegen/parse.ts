import { parseStruct, Decorator, ImportNode } from 'ts-file-parser';
import * as fs from 'fs';
// import * as path from 'path';
// import { ArrayType, BasicType } from 'ts-file-parser';

export interface DecoratedClassFieldMetadata {
  name: string;
  type: any;
  optional: any;
  decorators: Decorator[];
}

export interface DecoratedClassMetadata {
  name: string;
  decorators: Decorator[];
  fields: DecoratedClassFieldMetadata[];
}

export interface InputFileMetadata {
  name: string;
  imports: string[];
  classes: DecoratedClassMetadata[];
}

export const parseInputFiles = (files: string[]): InputFileMetadata[] => {
  return files.map((file) => {
    console.log(`Reading ${file}...`);
    const content = fs.readFileSync(file, 'utf-8');
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const structure = parseStruct(content, {}, file);

    //if (file.match(/^.*common.ts$/)) {
    //console.log(JSON.stringify(structure, null, 4));
    //}

    return {
      name: structure.name.replace(/\/+/g, '.'),
      imports: buildImports(structure._imports),
      classes: []
    };
  });
};

const buildImports = (importNodes: ImportNode[]): string[] => {
  return importNodes.map((iNode) => {
    return `import { ${iNode.clauses.join(', ')} } from ${iNode.absPathNode.join('/')}`;
  });
};
