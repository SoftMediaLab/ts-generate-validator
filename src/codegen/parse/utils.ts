import { primitiveValidationTypes } from './../../validators/model';
import { ValidationType } from '../../validators/model';

export const getValidationTypeByTypeName = (typeName: string): ValidationType => {
  if (primitiveValidationTypes.includes(typeName as any)) {
    return <ValidationType>typeName;
  }

  const notSupportedTypes = ['null', 'undefined', 'symbol', 'function', 'object'];
  if (notSupportedTypes.includes(typeName)) {
    return ValidationType.notSupported;
  }

  return ValidationType.unknown;
};
