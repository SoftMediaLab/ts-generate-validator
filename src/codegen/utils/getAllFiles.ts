import * as fs from 'fs';

export const getAllFiles = (path: string): string[] => {
  const tsRegExp = /.+\.ts$/;
  const returnFiles: string[] = [];

  const files = fs.readdirSync(path);
  files.forEach((file) => {
    const endPath = `${path}/${file}`;
    if (fs.statSync(endPath).isDirectory()) {
      const subFiles = getAllFiles(endPath);
      returnFiles.push(...subFiles);
    } else {
      const matches = tsRegExp.exec(endPath);
      if (matches && matches.length > 0) {
        returnFiles.push(matches[0]);
      }
    }
  });
  return returnFiles;
};
