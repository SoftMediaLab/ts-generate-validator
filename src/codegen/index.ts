import { parseInputFiles } from './parse';
import { getAllFiles } from './utils/getAllFiles';
import { getCodegenConfig } from './../config/codegen';

export const createValidators = (): void => {
  const config = getCodegenConfig();
  const inputFiles = getAllFiles(config.inputPath);
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const inputFilesMetadata = parseInputFiles(inputFiles);

  // TODO: implement
};
