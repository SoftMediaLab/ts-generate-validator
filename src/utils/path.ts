import { allowedFileExt } from './../codegen/model';
import * as path from 'path';
import * as fs from 'fs';

export const getRootDir = (): string => {
  return process.cwd();
};

export const isDir = (rawPath: string): boolean => {
  try {
    return fs.statSync(rawPath).isDirectory();
  } catch (err) {
    return false;
  }
};

export const cutFileExt = (file: string): string => {
  return file.replace(/\.[^/.]+$/, '');
};

export const hasFileExt = (file: string): boolean => {
  return /\.[^/.]+$/.test(file);
};

export const normalizeFileExt = (filePathAbs: string): string => {
  // TODO: use ts.resolveModuleName (ts.nodeModuleNameResolver) instead (refactor)
  let filePath = path.resolve(process.cwd(), filePathAbs);

  if (isDir(filePath)) {
    filePath = `${filePath}/index`;
  }

  if (isDir(filePath) || hasFileExt(filePath) || fs.existsSync(filePath)) {
    return filePathAbs;
  }

  for (const ext of allowedFileExt) {
    const possiblePath = `${filePath}.${ext}`;
    if (!isDir(possiblePath) && fs.existsSync(possiblePath)) {
      return path.relative(process.cwd(), possiblePath);
    }
  }

  return filePathAbs;
};

export const normalizePath = (rawPath: string): string => {
  return rawPath.replace(/\\+/g, '/');
};

export const isPackagePath = (rawPath: string): boolean => {
  return !hasFileExt(rawPath) && rawPath.indexOf('/') < 0;
};
