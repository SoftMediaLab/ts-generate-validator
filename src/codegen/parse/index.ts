import { cutFileExt, hasFileExt, isPackagePath, normalizeFileExt, normalizePath } from './../../utils/path';
import { IgnoreValidation } from './../../decorators/index';
import { IssueError, ErrorInFile, outWarning } from './../utils/error';
import { ValidationType } from './../../validators/model';
import { buildClassesMetadata } from './classes';
import { parseStruct, ImportNode } from 'ts-file-parser';
import * as fs from 'fs';
import * as path from 'path';
import {
  InputFileMetadata,
  ClassFieldTypeMetadata,
  ClassFieldBasicTypeMetadata,
  CustomTypeEntry,
  EnumDictionary,
  ImportMetadata,
  InputFileDesc,
  ClassMetadata
} from './model';

export const parseInputFiles = (files: string[]): InputFileMetadata[] => {
  const inputFiles: InputFileDesc[] = files.map((f) => ({ path: f, parseClasses: true }));
  const metadata: InputFileMetadata[] = [];
  const customTypeEntries: CustomTypeEntry[] = [];
  const enumDictionary: EnumDictionary = {};

  do {
    const inputFileDesc = inputFiles.shift();

    if (!inputFileDesc) {
      break;
    }

    const inputFilePath = path.resolve(process.cwd(), normalizeFileExt(inputFileDesc.path));
    let content: string;
    try {
      content = fs.readFileSync(inputFilePath, 'utf-8');
    } catch (err) {
      outWarning(`Failed to read input file "${inputFilePath}" -> skip (${err.message})`);
      continue;
    }
    const structure = parseStruct(content, {}, inputFileDesc.path);

    if (structure.enumDeclarations.length) {
      enumDictionary[inputFileDesc.path] = structure.enumDeclarations.map((decl) => decl.name);
    }

    const imports = buildImportsMetadata(structure._imports);
    imports.forEach((imp) => {
      const isAlreadyParsed = metadata.some((m) => m.name === imp.absPath);
      if (inputFileDesc.parseClasses && !isAlreadyParsed && !isPackagePath(imp.absPath)) {
        inputFiles.push({ path: imp.absPath, parseClasses: false });
      }
    });

    const classes: ClassMetadata[] = [];
    if (inputFileDesc.parseClasses && structure.classes.length) {
      const classesMetadata = buildClassesMetadata(metadata.length, structure.classes, structure._imports);
      if (classesMetadata.classesForValidation.length) {
        customTypeEntries.push(...classesMetadata.customTypeEntries);
        classes.push(...classesMetadata.classesForValidation);
      }
    }

    metadata.push({
      name: structure.name,
      classes,
      imports,
      functions: structure.functions.map(({ name, isExport, isAsync }) => ({ name, isExported: isExport, isAsync }))
    });
  } while (inputFiles.length);

  resolveCustomTypes({ metadata, enumDictionary, customTypeEntries });
  validateNestedClasses({ metadata, customTypeEntries });

  return metadata;
};

export const resolveCustomTypes = ({
  metadata,
  enumDictionary,
  customTypeEntries
}: {
  metadata: InputFileMetadata[];
  enumDictionary: EnumDictionary;
  customTypeEntries: CustomTypeEntry[];
}): void => {
  customTypeEntries.forEach(({ fileIndex, classIndex, fieldIndex }) => {
    const fieldTypeMetadata = metadata[fileIndex].classes[classIndex].fields[fieldIndex].type;

    const resolveBasicType = (typeMetadata: ClassFieldBasicTypeMetadata): void => {
      if (typeMetadata.validationType !== ValidationType.unknown) {
        return;
      }

      // Trying to fix empty referencePath -> set to current file with model class
      if (!typeMetadata.referencePath) {
        typeMetadata.referencePath = metadata[fileIndex].name;
      }

      const { referencePath, name: typeName } = typeMetadata;
      if (!referencePath || !typeName) {
        typeMetadata.validationType = ValidationType.notSupported;
        return;
      }

      if (enumDictionary[referencePath]?.includes(typeName)) {
        typeMetadata.validationType = ValidationType.enum;
        return;
      }

      const metadataItemWithNestedClass = metadata.find(({ name: fileName, classes }) => {
        let fileNameMatched = true;

        const fileNameHasExt = hasFileExt(fileName);
        const refPathHasExt = hasFileExt(referencePath);

        if (fileNameHasExt === refPathHasExt) {
          fileNameMatched = fileName === referencePath;
        } else {
          fileNameMatched = cutFileExt(fileName) === cutFileExt(referencePath);
        }

        if (!fileNameMatched) {
          return false;
        }

        return Boolean(classes.find(({ name: className }) => className === typeName));
      });
      if (metadataItemWithNestedClass) {
        typeMetadata.validationType = ValidationType.nested;
        typeMetadata.referencePath = metadataItemWithNestedClass.name;
        return;
      }

      typeMetadata.validationType = ValidationType.notSupported;
    };

    const resolveAnyType = (typeMetadata: ClassFieldTypeMetadata): void => {
      if (typeMetadata.validationType === ValidationType.array) {
        return resolveAnyType(typeMetadata.arrayOf);
      }
      if (typeMetadata.validationType === ValidationType.union) {
        return typeMetadata.unionTypes.forEach(resolveAnyType);
      }
      return resolveBasicType(typeMetadata);
    };

    resolveAnyType(fieldTypeMetadata);
  });
};

export const validateNestedClasses = ({
  metadata,
  customTypeEntries
}: {
  metadata: InputFileMetadata[];
  customTypeEntries: CustomTypeEntry[];
}): void => {
  customTypeEntries.forEach(({ fileIndex, classIndex, fieldIndex }) => {
    const baseClassName = metadata[fileIndex].classes[classIndex].name;
    const baseClassPath = metadata[fileIndex].name;
    const fieldTypeMetadata = metadata[fileIndex].classes[classIndex].fields[fieldIndex].type;

    const checkBasicType = (typeMetadata: ClassFieldBasicTypeMetadata): void => {
      if (typeMetadata.validationType !== ValidationType.nested) {
        return;
      }

      const checkNestedClasses = (className: string, classPath: string) => {
        if (className === baseClassName && classPath === baseClassPath) {
          const fieldName = metadata[fileIndex].classes[classIndex].fields[fieldIndex].name;
          throw new ErrorInFile(
            `Class "${baseClassName}" has circular dependency in field "${fieldName}" with type "${typeMetadata.name}". Ð¡ircular dependencies are not allowed because they lead to infinite validation. Change field type or add "@${IgnoreValidation.name}" decorator.`,
            baseClassPath
          );
        }

        const _fileMetadata = metadata.find(({ name: filePath }) => filePath === classPath);
        if (!_fileMetadata) {
          throw new IssueError(
            `Referenced file metadata not found for nested class "${className}" (referenced file path: "${classPath}").`
          );
        }

        const _classMetadata = _fileMetadata.classes.find(({ name }) => name === className);
        if (!_classMetadata) {
          throw new IssueError(`Metadata not found for nested class "${className}" in file metadata "${classPath}").`);
        }

        _classMetadata.fields.forEach(({ type: _fieldTypeMetadata }) => {
          const _checkBasicType = (_typeMetadata: ClassFieldBasicTypeMetadata): void => {
            const { name: _name, validationType: _validationType, referencePath: _referencePath } = _typeMetadata;

            if (_validationType === ValidationType.nested && _name && _referencePath) {
              checkNestedClasses(_name, _referencePath);
            }
          };

          const _checkAnyType = (_typeMetadata: ClassFieldTypeMetadata): void => {
            if (_typeMetadata.validationType === ValidationType.array) {
              return _checkAnyType(_typeMetadata.arrayOf);
            }
            if (_typeMetadata.validationType === ValidationType.union) {
              return _typeMetadata.unionTypes.forEach(_checkAnyType);
            }
            return _checkBasicType(_typeMetadata);
          };

          _checkAnyType(_fieldTypeMetadata);
        });
      };

      if (typeMetadata.name && typeMetadata.referencePath) {
        checkNestedClasses(typeMetadata.name, typeMetadata.referencePath);
      }
    };

    const checkAnyType = (typeMetadata: ClassFieldTypeMetadata): void => {
      if (typeMetadata.validationType === ValidationType.array) {
        return checkAnyType(typeMetadata.arrayOf);
      }
      if (typeMetadata.validationType === ValidationType.union) {
        return typeMetadata.unionTypes.forEach(checkAnyType);
      }
      return checkBasicType(typeMetadata);
    };

    checkAnyType(fieldTypeMetadata);
  });
};

const buildImportsMetadata = (nodes: ImportNode[]): ImportMetadata[] => {
  return nodes.map(({ clauses, absPathString }) => {
    return {
      clauses,
      absPath: normalizePath(path.relative(process.cwd(), path.resolve(absPathString)))
    };
  });
};
