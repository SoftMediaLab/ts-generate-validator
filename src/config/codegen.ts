import { outWarning } from './../codegen/utils/error';
import { getSanitizedObjectCopy, isObject } from './../utils/object';
import { getEnumValues } from './../utils/enum';
import { CodegenConfig, PartialCodegenConfig, SeverityLevel, codegenConfigKeys } from './model';
import { defaultCodegenConfig } from './default';
import { mergeDeep } from './../utils/deepValue';
import * as fs from 'fs';
import * as path from 'path';

export const configFileName = 'ts-generate-validator-config.json';

const getConfigFromFile = (): Record<string, any> => {
  const configFilePath = path.resolve(process.cwd(), configFileName);

  if (!fs.existsSync(configFilePath)) {
    return {};
  }
  try {
    const parsedData = JSON.parse(fs.readFileSync(configFilePath, 'utf8'));

    if (!isObject(parsedData)) {
      throw new Error();
    }

    return parsedData;
  } catch (err) {
    outWarning(
      `Error when reading config from file "${configFileName}". Make sure the file has correct syntax. Using default config...`
    );
    return {};
  }
};

export const getCodegenConfig = (): CodegenConfig => {
  const configFromFile = getConfigFromFile();
  const sanitizedConfig = getSanitizedObjectCopy<PartialCodegenConfig>(configFromFile, codegenConfigKeys);

  if (sanitizedConfig.hasOwnProperty('inputPath')) {
    if (typeof sanitizedConfig.inputPath !== 'string') {
      outWarning(
        `File "${configFileName}" has wrong configuration. Property "inputPath" is not a valid path. Using default value "${defaultCodegenConfig.inputPath}".`
      );
      delete sanitizedConfig.inputPath;
    }
  }

  if (sanitizedConfig.hasOwnProperty('outputPath')) {
    if (typeof sanitizedConfig.outputPath !== 'string') {
      outWarning(
        `File "${configFileName}" has wrong configuration. Property "outputPath" is not a valid path. Using default value "${defaultCodegenConfig.outputPath}".`
      );
      delete sanitizedConfig.inputPath;
    }
  }

  if (sanitizedConfig.hasOwnProperty('unknownPropertySeverityLevel')) {
    const possibleValues = getEnumValues(SeverityLevel);
    if (!possibleValues.includes(String(sanitizedConfig.unknownPropertySeverityLevel))) {
      outWarning(
        `File "${configFileName}" has wrong configuration. Property "unknownPropertySeverityLevel" must be one of "${possibleValues.join(
          ', '
        )}", but recieved "${sanitizedConfig.unknownPropertySeverityLevel}". Using default value "${
          defaultCodegenConfig.unknownPropertySeverityLevel
        }".`
      );
      delete sanitizedConfig.unknownPropertySeverityLevel;
    }
  }

  return mergeDeep({}, defaultCodegenConfig, sanitizedConfig) as CodegenConfig;
};
