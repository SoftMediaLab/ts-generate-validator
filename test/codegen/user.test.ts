import { parseInputFiles } from './../../src/codegen/parse/index';
import * as path from 'path';
import { getAllFiles } from './../../src/codegen/utils/getAllFiles';
import { getCodegenConfig } from './../../src/config/codegen';
import { deleteConfigFile, createConfigFile } from './../utils/config';

describe('codegen/user', () => {
  const prevRootDir = process.cwd();

  beforeAll(() => {
    const testRootDir = process.cwd() + '/test/mock/user';
    process.chdir(testRootDir);
    createConfigFile({
      inputPath: '.'
    });
  });

  afterAll(() => {
    deleteConfigFile();
    process.chdir(prevRootDir);
  });

  test('parse metadata', () => {
    const config = getCodegenConfig();
    const inputFiles = getAllFiles(path.resolve(config.inputPath));
    const inputFilesMetadata = parseInputFiles(inputFiles);

    expect(inputFilesMetadata).toMatchSnapshot();
  });
});
