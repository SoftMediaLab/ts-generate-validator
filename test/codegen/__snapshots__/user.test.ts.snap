// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen/user create validators: generated validators at "/Users/alex-shul/Desktop/projects/ts-generate-validator/test/mock/user/generated/validators/model.ts" 1`] = `
"import {
  GeneratedValidation,
  GeneratedValidationPayload,
  UserContext,
  initConfig,
  getConfig,
  mergeDeep,
  requiredOneOfValidator,
  typeValidator,
  minLengthValidator,
  maxLengthValidator,
  trimValidator,
  minValidator,
  maxValidator,
  integerValidator,
  lessThanValidator,
  positiveValidator,
  equalValidator,
  emailValidator,
  matchValidator,
  urlValidator,
  lowercaseValidator,
  uppercaseValidator,
  floatValidator
} from 'ts-generate-validator';

import { User } from '../../model';

export const userValidator: GeneratedValidation = (payload: GeneratedValidationPayload<typeof User, UserContext>) => {
  const config = mergeDeep({}, getConfig(), payload.config);

  requiredOneOfValidator({
    property: payload.data.email,
    propertyName: 'email',
    data: payload.data,
    config,

    fields: ['email', 'phone']
  });

  typeValidator({
    property: payload.data.name,
    propertyName: 'name',
    data: payload.data,
    config
  });

  minLengthValidator({
    property: payload.data.name,
    propertyName: 'name',
    data: payload.data,
    config,

    threshold: 2
  });

  maxLengthValidator({
    property: payload.data.name,
    propertyName: 'name',
    data: payload.data,
    config,

    threshold: 50
  });

  trimValidator({
    property: payload.data.name,
    propertyName: 'name',
    data: payload.data,
    config
  });

  typeValidator({
    property: payload.data.surname,
    propertyName: 'surname',
    data: payload.data,
    config
  });

  minLengthValidator({
    property: payload.data.surname,
    propertyName: 'surname',
    data: payload.data,
    config,

    threshold: 2,

    customMessage: 'surname length must be more then 2 letters'
  });

  maxLengthValidator({
    property: payload.data.surname,
    propertyName: 'surname',
    data: payload.data,
    config,

    threshold: 50,

    customMessage: 'surname length must be less then 50 letters'
  });

  trimValidator({
    property: payload.data.surname,
    propertyName: 'surname',
    data: payload.data,
    config,

    customMessage: 'there should be no spaces at the beginning and at the end of surname'
  });

  typeValidator({
    property: payload.data.age,
    propertyName: 'age',
    data: payload.data,
    config
  });

  minValidator({
    property: payload.data.age,
    propertyName: 'age',
    data: payload.data,
    config,

    threshold: 18
  });

  maxValidator({
    property: payload.data.age,
    propertyName: 'age',
    data: payload.data,
    config,

    threshold: 60
  });

  integerValidator({
    property: payload.data.age,
    propertyName: 'age',
    data: payload.data,
    config
  });

  typeValidator({
    property: payload.data.workExperience,
    propertyName: 'workExperience',
    data: payload.data,
    config
  });

  lessThanValidator({
    property: payload.data.workExperience,
    propertyName: 'workExperience',
    data: payload.data,
    config,

    targetPropertyName: 'age'
  });

  positiveValidator({
    property: payload.data.workExperience,
    propertyName: 'workExperience',
    data: payload.data,
    config
  });

  typeValidator({
    property: payload.data.isActive,
    propertyName: 'isActive',
    data: payload.data,
    config
  });

  equalValidator({
    property: payload.data.isActive,
    propertyName: 'isActive',
    data: payload.data,
    config,

    value: true
  });

  typeValidator({
    property: payload.data.email,
    propertyName: 'email',
    data: payload.data,
    config
  });

  emailValidator({
    property: payload.data.email,
    propertyName: 'email',
    data: payload.data,
    config
  });

  typeValidator({
    property: payload.data.phone,
    propertyName: 'phone',
    data: payload.data,
    config
  });

  matchValidator({
    property: payload.data.phone,
    propertyName: 'phone',
    data: payload.data,
    config,

    regexp: /^\\\\d{10}$/
  });

  typeValidator({
    property: payload.data.website,
    propertyName: 'website',
    data: payload.data,
    config
  });

  urlValidator({
    property: payload.data.website,
    propertyName: 'website',
    data: payload.data,
    config
  });

  typeValidator({
    property: payload.data.nickname,
    propertyName: 'nickname',
    data: payload.data,
    config
  });

  lowercaseValidator({
    property: payload.data.nickname,
    propertyName: 'nickname',
    data: payload.data,
    config
  });

  minLengthValidator({
    property: payload.data.nickname,
    propertyName: 'nickname',
    data: payload.data,
    config,

    threshold: 5
  });

  maxLengthValidator({
    property: payload.data.nickname,
    propertyName: 'nickname',
    data: payload.data,
    config,

    threshold: 20
  });

  typeValidator({
    property: payload.data.cardHolderName,
    propertyName: 'cardHolderName',
    data: payload.data,
    config
  });

  uppercaseValidator({
    property: payload.data.cardHolderName,
    propertyName: 'cardHolderName',
    data: payload.data,
    config
  });

  typeValidator({
    property: payload.data.insuranceRatio,
    propertyName: 'insuranceRatio',
    data: payload.data,
    config
  });

  floatValidator({
    property: payload.data.insuranceRatio,
    propertyName: 'insuranceRatio',
    data: payload.data,
    config
  });
};
"
`;

exports[`codegen/user parse & prepare: data for render 1`] = `
Array [
  Object {
    "configFilePath": undefined,
    "fileName": "model",
    "filePath": "/Users/alex-shul/Desktop/projects/ts-generate-validator/test/mock/user/generated/validators",
    "imports": Array [
      Object {
        "clauses": "GeneratedValidation, GeneratedValidationPayload, UserContext, initConfig, getConfig, mergeDeep, requiredOneOfValidator, typeValidator, minLengthValidator, maxLengthValidator, trimValidator, minValidator, maxValidator, integerValidator, lessThanValidator, positiveValidator, equalValidator, emailValidator, matchValidator, urlValidator, lowercaseValidator, uppercaseValidator, floatValidator",
        "path": "ts-generate-validator",
      },
      Object {
        "clauses": "User",
        "path": "../../model",
      },
    ],
    "validationArgs": Object {
      "config": "config",
      "data": "data",
    },
    "validations": Array [
      Object {
        "async": false,
        "items": Array [
          Object {
            "propertyName": "email",
            "validatorName": "requiredOneOfValidator",
            "validatorPayload": Array [
              Object {
                "property": "fields",
                "type": "array",
                "value": "['email', 'phone']",
              },
            ],
          },
          Object {
            "propertyName": "name",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "name",
            "validatorName": "minLengthValidator",
            "validatorPayload": Array [
              Object {
                "property": "threshold",
                "type": "",
                "value": "2",
              },
            ],
          },
          Object {
            "propertyName": "name",
            "validatorName": "maxLengthValidator",
            "validatorPayload": Array [
              Object {
                "property": "threshold",
                "type": "",
                "value": "50",
              },
            ],
          },
          Object {
            "propertyName": "name",
            "validatorName": "trimValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "surname",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "surname",
            "validatorName": "minLengthValidator",
            "validatorPayload": Array [
              Object {
                "property": "threshold",
                "type": "",
                "value": "2",
              },
              Object {
                "property": "customMessage",
                "type": "string",
                "value": "surname length must be more then 2 letters",
              },
            ],
          },
          Object {
            "propertyName": "surname",
            "validatorName": "maxLengthValidator",
            "validatorPayload": Array [
              Object {
                "property": "threshold",
                "type": "",
                "value": "50",
              },
              Object {
                "property": "customMessage",
                "type": "string",
                "value": "surname length must be less then 50 letters",
              },
            ],
          },
          Object {
            "propertyName": "surname",
            "validatorName": "trimValidator",
            "validatorPayload": Array [
              Object {
                "property": "customMessage",
                "type": "string",
                "value": "there should be no spaces at the beginning and at the end of surname",
              },
            ],
          },
          Object {
            "propertyName": "age",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "age",
            "validatorName": "minValidator",
            "validatorPayload": Array [
              Object {
                "property": "threshold",
                "type": "",
                "value": "18",
              },
            ],
          },
          Object {
            "propertyName": "age",
            "validatorName": "maxValidator",
            "validatorPayload": Array [
              Object {
                "property": "threshold",
                "type": "",
                "value": "60",
              },
            ],
          },
          Object {
            "propertyName": "age",
            "validatorName": "integerValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "workExperience",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "workExperience",
            "validatorName": "lessThanValidator",
            "validatorPayload": Array [
              Object {
                "property": "targetPropertyName",
                "type": "string",
                "value": "age",
              },
            ],
          },
          Object {
            "propertyName": "workExperience",
            "validatorName": "positiveValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "isActive",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "isActive",
            "validatorName": "equalValidator",
            "validatorPayload": Array [
              Object {
                "property": "value",
                "type": "boolean",
                "value": "true",
              },
            ],
          },
          Object {
            "propertyName": "email",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "email",
            "validatorName": "emailValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "phone",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "phone",
            "validatorName": "matchValidator",
            "validatorPayload": Array [
              Object {
                "property": "regexp",
                "type": "",
                "value": "/^\\\\d{10}$/",
              },
            ],
          },
          Object {
            "propertyName": "website",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "website",
            "validatorName": "urlValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "nickname",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "nickname",
            "validatorName": "lowercaseValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "nickname",
            "validatorName": "minLengthValidator",
            "validatorPayload": Array [
              Object {
                "property": "threshold",
                "type": "",
                "value": "5",
              },
            ],
          },
          Object {
            "propertyName": "nickname",
            "validatorName": "maxLengthValidator",
            "validatorPayload": Array [
              Object {
                "property": "threshold",
                "type": "",
                "value": "20",
              },
            ],
          },
          Object {
            "propertyName": "cardHolderName",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "cardHolderName",
            "validatorName": "uppercaseValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "insuranceRatio",
            "validatorName": "typeValidator",
            "validatorPayload": Array [],
          },
          Object {
            "propertyName": "insuranceRatio",
            "validatorName": "floatValidator",
            "validatorPayload": Array [],
          },
        ],
        "modelClassName": "User",
        "name": "userValidator",
      },
    ],
  },
]
`;

exports[`codegen/user parse & prepare: parsed metadata 1`] = `
Array [
  Object {
    "classes": Array [
      Object {
        "decorators": Array [
          Object {
            "arguments": Array [],
            "name": "Validation",
          },
          Object {
            "arguments": Array [
              Array [
                "email",
                "phone",
              ],
            ],
            "name": "RequiredOneOfValidation",
          },
        ],
        "fields": Array [
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  2,
                ],
                "name": "MinLengthValidation",
              },
              Object {
                "arguments": Array [
                  50,
                ],
                "name": "MaxLengthValidation",
              },
              Object {
                "arguments": Array [],
                "name": "TrimValidation",
              },
            ],
            "name": "name",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  2,
                  "surname length must be more then 2 letters",
                ],
                "name": "MinLengthValidation",
              },
              Object {
                "arguments": Array [
                  50,
                  "surname length must be less then 50 letters",
                ],
                "name": "MaxLengthValidation",
              },
              Object {
                "arguments": Array [
                  "there should be no spaces at the beginning and at the end of surname",
                ],
                "name": "TrimValidation",
              },
            ],
            "name": "surname",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  18,
                ],
                "name": "MinValidation",
              },
              Object {
                "arguments": Array [
                  60,
                ],
                "name": "MaxValidation",
              },
              Object {
                "arguments": Array [],
                "name": "IntegerValidation",
              },
            ],
            "name": "age",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "age",
                ],
                "name": "LessThanValidation",
              },
              Object {
                "arguments": Array [],
                "name": "PositiveValidation",
              },
            ],
            "name": "workExperience",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  true,
                ],
                "name": "EqualValidation",
              },
            ],
            "name": "isActive",
            "optional": false,
            "type": Object {
              "name": "boolean",
              "validationType": "boolean",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "EmailValidation",
              },
            ],
            "name": "email",
            "optional": true,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [
                  "/^\\\\d{10}$/",
                ],
                "name": "MatchValidation",
              },
            ],
            "name": "phone",
            "optional": true,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "UrlValidation",
              },
            ],
            "name": "website",
            "optional": true,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "LowercaseValidation",
              },
              Object {
                "arguments": Array [
                  5,
                ],
                "name": "MinLengthValidation",
              },
              Object {
                "arguments": Array [
                  20,
                ],
                "name": "MaxLengthValidation",
              },
            ],
            "name": "nickname",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "UppercaseValidation",
              },
            ],
            "name": "cardHolderName",
            "optional": false,
            "type": Object {
              "name": "string",
              "validationType": "string",
            },
          },
          Object {
            "decorators": Array [
              Object {
                "arguments": Array [],
                "name": "FloatValidation",
              },
            ],
            "name": "insuranceRatio",
            "optional": false,
            "type": Object {
              "name": "number",
              "validationType": "number",
            },
          },
          Object {
            "decorators": Array [],
            "name": "userType",
            "optional": true,
            "type": Object {
              "name": "UserType",
              "referencePath": "/Users/alex-shul/Desktop/projects/ts-generate-validator/test/mock/user/model",
              "validationType": "notSupported",
            },
          },
        ],
        "name": "User",
      },
    ],
    "name": "/Users/alex-shul/Desktop/projects/ts-generate-validator/test/mock/user/model",
  },
]
`;
